### PR TITLE
fix(autorag+automl): Exclude zero-byte folder marker objects in S3 list files BFF

### PR DESCRIPTION
**Pull request by GAUNSD**
## Issues

- https://redhat.atlassian.net/browse/RHOAIENG-62235

## Description

This PR addresses and issue with S3FileExplorer rendering zero-byte file markers that some S3 tools explicitly create when managing s3-virtual folders. S3 connections in the past did not have these zero-byte markers and recent tests show this bug.

|Before|After|
|--|--|
|Untitled|image|

## How Has This Been Tested?

- Ran autorag go BFF manually with the S3FileExplorer playground

## Test Impact

- Added go BFF tests for the behaviour change in the `client_test.go`

## Request review criteria:

Self checklist (all need to be checked):
- [X] The developer has manually tested the changes and verified that the changes work
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)
- [x] The code follows our [Best Practices](/docs/best-practices.md) (React coding standards, PatternFly usage, performance considerations)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

## Summary by CodeRabbit

* **Bug Fixes**
  * Filter out zero-byte “folder marker” objects so a folder no longer appears as its own child in listings.
  * Preserve real files (including those with spaces in paths) and their sizes when listing directory contents.
  * Ensure root-level listings are not incorrectly filtered, and directories containing only their marker return empty contents.

### CODE REVIEW

File: packages/autorag/bff/internal/integrations/s3/client.go [hunk 1]
```diff
@@ -252,8 +252,16 @@ func (c *RealS3Client) ListObjects(ctx context.Context, bucket string, options L
...
 
 	for _, obj := range output.Contents {
+		key := aws.ToString(obj.Key)
+		// Skip the folder marker that represents the queried prefix itself.
+		// S3 returns zero-byte objects whose key equals the prefix (e.g.
+		// "folder/") as content items; surfacing them causes the UI to
+		// render the current folder as a child of itself.
+		if key == prefix {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (GAUNSD) [reply packages/autorag/bff/internal/integrations/s3/client.go:261]: Pushed a fix: https://github.com/opendatahub-io/odh-dashboard/pull/7535/commits/c9f78eaa1298384f4afba15e70a529b1248fd815

File: packages/automl/bff/internal/integrations/s3/client.go [hunk 2]
```diff
@@ -291,8 +291,16 @@ func (c *RealS3Client) ListObjects(ctx context.Context, bucket string, options L
...
 
 	for _, obj := range output.Contents {
+		key := aws.ToString(obj.Key)
+		// Skip the folder marker that represents the queried prefix itself.
+		// S3 returns zero-byte objects whose key equals the prefix (e.g.
+		// "folder/") as content items; surfacing them causes the UI to
+		// render the current folder as a child of itself.
+		if key == prefix {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (GAUNSD) [reply packages/automl/bff/internal/integrations/s3/client.go:310]: Pushed a fix: https://github.com/opendatahub-io/odh-dashboard/pull/7535/commits/c9f78eaa1298384f4afba15e70a529b1248fd815

**Comment by nickmazzi** [comment]
Thanks @GAUNSD 
Works great. Don’t think any of the suggestions are necessary. Will approve.

# AI assisted review
[pr-review-7535.md](https://github.com/user-attachments/files/28277626/pr-review-7535.md)

## Overall Assessment

**Approve** — Clean, correct bug fix with excellent test coverage. A few minor observations below.

## Suggestions

~1. **Future consolidation**: The autorag and automl S3 clients are nearly identical. Consider extracting a shared S3 client package to avoid maintaining identical code in two places. Not blocking for this PR.~

2. **Edge case — non-zero-byte "folder" keys**: Some S3 tools create folder markers with a small body (e.g., 1 byte). The current filter only catches zero-byte markers. This is fine for now since the standard convention is zero-byte, but worth documenting if edge cases appear later.

3. **Test parallelism in table-driven test**: In `TestListObjects_SkipsFolderMarker`, the subtests created via `t.Run` inside `for _, tt := range tests` don't capture `tt` in a local variable. In Go 1.22+ this is fine (loop var scoping changed), but if this repo's Go toolchain is <1.22, this could be a subtle bug. The `go.mod` should confirm — given the repo requires Go >= 1.24 per CLAUDE.md, this is safe.


# Test Validation

## autorag

### Frontend ✅
```
Test Suites: 48 passed, 48 total
Tests:       897 passed, 897 total
Snapshots:   0 total
Time:        22.397 s
```

### BFF ✅
```
go fmt ./...
go vet ./...
ok  	github.com/opendatahub-io/autorag-library/bff/cmd	(cached)
ok  	github.com/opendatahub-io/autorag-library/bff/internal/api	(cached)
?   	github.com/opendatahub-io/autorag-library/bff/internal/config	[no test files]
?   	github.com/opendatahub-io/autorag-library/bff/internal/constants	[no test files]
?   	github.com/opendatahub-io/autorag-library/bff/internal/helpers	[no test files]
?   	github.com/opendatahub-io/autorag-library/bff/internal/integrations	[no test files]
ok  	github.com/opendatahub-io/autorag-library/bff/internal/integrations/kubernetes	(cached)
?   	github.com/opendatahub-io/autorag-library/bff/internal/integrations/kubernetes/k8mocks	[no test files]
?   	github.com/opendatahub-io/autorag-library/bff/internal/integrations/kubernetes/k8smocks	[no test files]
ok  	github.com/opendatahub-io/autorag-library/bff/internal/integrations/ogx	(cached)
?   	github.com/opendatahub-io/autorag-library/bff/internal/integrations/ogx/ogxmocks	[no test files]
ok  	github.com/opendatahub-io/autorag-library/bff/internal/integrations/pipelineserver	(cached)
?   	github.com/opendatahub-io/autorag-library/bff/internal/integrations/pipelineserver/psmocks	[no test files]
ok  	github.com/opendatahub-io/autorag-library/bff/internal/integrations/s3	0.596s
?   	github.com/opendatahub-io/autorag-library/bff/internal/integrations/s3/s3mocks	[no test files]
?   	github.com/opendatahub-io/autorag-library/bff/internal/mocks	[no test files]
?   	github.com/opendatahub-io/autorag-library/bff/internal/models	[no test files]
?   	github.com/opendatahub-io/autorag-library/bff/internal/pipelines	[no test files]
ok  	github.com/opendatahub-io/autorag-library/bff/internal/repositories	(cached)
```

### Contract ✅
```
Test Suites: 1 passed, 1 total
Tests:       111 passed, 111 total
Snapshots:   0 total
Time:        2.08 s
Test Suites: 1 passed, 1 total
Tests:       111 passed, 111 total
```

### BFF Lint ✅
```
No lint issues
```

### Lint ✅
```
No lint output
```

### Type Check ✅
```
No type errors
```

---

### Overall Summary

| Suite      | Result |
|------------|--------|
| Frontend   | ✅     |
| BFF        | ✅     |
| Contract   | ✅     |
| BFF Lint   | ✅     |
| Lint       | ✅     |
| Type Check | ✅     |

## automl

### Frontend ✅
```
Test Suites: 53 passed, 53 total
Tests:       1031 passed, 1031 total
Snapshots:   0 total
Time:        11.627 s, estimated 25 s
```

### BFF ✅
```
go fmt ./...
go vet ./...
ok  	github.com/opendatahub-io/automl-library/bff/cmd	(cached)
ok  	github.com/opendatahub-io/automl-library/bff/internal/api	(cached)
?   	github.com/opendatahub-io/automl-library/bff/internal/config	[no test files]
?   	github.com/opendatahub-io/automl-library/bff/internal/constants	[no test files]
?   	github.com/opendatahub-io/automl-library/bff/internal/helpers	[no test files]
?   	github.com/opendatahub-io/automl-library/bff/internal/integrations	[no test files]
ok  	github.com/opendatahub-io/automl-library/bff/internal/integrations/kubernetes	(cached)
?   	github.com/opendatahub-io/automl-library/bff/internal/integrations/kubernetes/k8mocks	[no test files]
?   	github.com/opendatahub-io/automl-library/bff/internal/integrations/modelregistry	[no test files]
ok  	github.com/opendatahub-io/automl-library/bff/internal/integrations/pipelineserver	(cached)
?   	github.com/opendatahub-io/automl-library/bff/internal/integrations/pipelineserver/psmocks	[no test files]
ok  	github.com/opendatahub-io/automl-library/bff/internal/integrations/s3	0.666s
?   	github.com/opendatahub-io/automl-library/bff/internal/integrations/s3/s3mocks	[no test files]
?   	github.com/opendatahub-io/automl-library/bff/internal/mocks	[no test files]
?   	github.com/opendatahub-io/automl-library/bff/internal/models	[no test files]
?   	github.com/opendatahub-io/automl-library/bff/internal/pipelines	[no test files]
ok  	github.com/opendatahub-io/automl-library/bff/internal/repositories	(cached)
```

### Contract ✅
```
Test Suites: 1 passed, 1 total
Tests:       94 passed, 94 total
Snapshots:   0 total
Time:        0.859 s, estimated 2 s
Test Suites: 1 passed, 1 total
Tests:       94 passed, 94 total
```

### BFF Lint ✅
```
No lint issues
```

### Lint ✅
```
No lint output
```

### Type Check ✅
```
No type errors
```

---

### Overall Summary

| Suite      | Result |
|------------|--------|
| Frontend   | ✅     |
| BFF        | ✅     |
| Contract   | ✅     |
| BFF Lint   | ✅     |
| Lint       | ✅     |
| Type Check | ✅     |

# Manual testing

<table><Tr><Td>tool</td><td>before</td><td>after</td></tr>
<tr><td>rag</td><td>
image

</td><td>

image
</td></tr>

<tr><td>ml</td><td>
image

</td><td>
image

</td></tr></table>

Also tested an edge case where I named a child folder the same as its parent and the fix does not remove the nested folder:

<table><tr><td>before</td><Td>after</td></tr>
<tr><td>
image

</td><td>
image
image

</td></tr></table>
